PLMS.BLL/ServicesImplementation/TaskService.cs

```diff
@@ -1,206 +1,204 @@
+using PLMS.BLL.DTO;
 using PLMS.BLL.Filters;
+using PLMS.BLL.ServicesInterfaces;
+using PLMS.Common.Exceptions;
 using PLMS.DAL.Interfaces;
+using System.Linq.Expressions;
 using Task = PLMS.DAL.Entities.Task;
 
 namespace PLMS.BLL.ServicesImplementation
 {
     public class TaskService : ITaskService
     {
-
         private readonly IUnitOfWork _unitOfWork;
 
+        private readonly IMapper _mapper;
         private readonly IRepository<Task> _taskRepository;
 
-        public TaskService(IUnitOfWork unitOfWork)
+        public TaskService(IUnitOfWork unitOfWork, IMapper mapper)
         {
             _unitOfWork = unitOfWork;
+            _mapper = mapper;
             _taskRepository = _unitOfWork.GetRepository<Task>();
         }
-        public async System.Threading.Tasks.Task AddTaskAsync(TaskDto taskDto)//ToDo: should be a separate DTO class for adding Task
+
+        public async Task<int> AddTaskAsync(AddTaskDto taskDto)
         {
-            Task task = new()
-            {
-                Title = taskDto.Title,
-                Description = taskDto.Description,
-                //GoalId = taskDto.GoalId,//you can use this for linking existing Goal
-                //Goal = taskDto.GoalDto, //you can use this for creating a new Goal, don't forget to map it from DTO to EF entity
-                //Category = taskDto.CategoryDto,//same
-                //Priority = taskDto.PriorityDto//same
-                CreatedAt = taskDto.CreatedAt,//TODO: User should not be able to set this field, it should not be present in the DTO, it should be autofilled here instead
-                DueDate = taskDto.DueDate,
-                TaskComments = taskDto.Comments.Select(x => new TaskComment { Comment = x.Comment, CreatedAt = x.CreatedAt}).ToArray()
-            };
+            var task = _mapper.Map<Task>(taskDto);
             await _taskRepository.CreateAsync(task);
             await _unitOfWork.CommitChangesToDatabaseAsync();
+
+            return task.Id;
         }
 
-        public async System.Threading.Tasks.Task DeleteTaskAsync(int id)
+        public async System.Threading.Tasks.Task DeleteTaskAsync(int id, string userId)
         {
-            Task task = await _taskRepository.GetByIdAsync(id);
+            Task task = await _taskRepository.GetByPredicateAsync(t => t.Id == id, t => t.Goal);
 
             if (task == null)
             {
                 return;
             }
 
+            if (task.Goal.UserId != userId)
+            {
+                throw new UnauthorizedAccessException("This user is not unauthorized to delete this task");
+            }
+
             _taskRepository.Remove(task);
             await _unitOfWork.CommitChangesToDatabaseAsync();
         }
 
-        public async System.Threading.Tasks.Task EditTaskAsync(TaskDto taskDto)//ToDo: should be a separate DTO class for editing Task
+        public async System.Threading.Tasks.Task EditTaskAsync(EditTaskDto taskDto, int id, string userId)
```

Comment:
move id into EditTaskDto class

Reply:
If TaskBaseDto includes an Id, it gets inherited by AddTaskDto, which shouldn't have an Id. Since you're creating a new entity, you should not accept an Id in the Add method. The parameter of the Add method should explicitly express this contract. 
So, please move the Id from TaskBaseDto to EditTaskDto.

Reply:
right, I didn't take it into account.